### PR TITLE
Treat null/nil server capabilities as false

### PR DESCRIPTION
**Pull request by jorams**
Some language servers (the PHP one for example) may specify null for
some capabilities in the list of server capabilities. This does not
conform to the specification, but treating it as false is more
reasonable than treating it as true.

### CODE REVIEW

**Comment by MaskRay** [comment]
Should the problem be reported in the language server repository?

**Comment by nemethf** [comment]
I agree it would be good to notify the servers' authors.

Additionally, I think the patch might not be always correct.  Eglot.el has this: `(eglot--server-capable :completionProvider :triggerCharacters)`, where the return value is expected to be a list and not a boolean value.  I'm not sure it's a good idea if eglot--server-capable returns false here instead of nil.

**Comment by joaotavora** [comment]
> I'm not sure it's a good idea if eglot--server-capable returns false here instead of nil.

Good, broad point: all the uses of `eglot--server-capable` must remain coherent. 

But I don't understand what is the php server's response vs a compliant server's response.  As such I don't understand what the consequences would be for each of those uses.  A couple of concrete examples would probably illustrate the issue.

**Comment by joaotavora** [comment]
Sorry, I closed this by mistake.

**Comment by nemethf** [comment]
In the mean time I realized what I wrote doesn't really make sense: the PR will correctly return nil if there are no triggerCharacters.

**Comment by joaotavora** [comment]
> what I wrote doesn't really make sense

It makes sense broadly, and was also my concern. But if you have double checked that this doesn't introduce new bugs in that function, then I don't oppose this.  Three very small but important things: 

* please  amend the commit to add a one-line comment near the particular line, explaining the decision;

* please amend the commit message to conform to the convention;

* please check if @jorams has FSF papers signed. If he/she doesn't, we need to add `Copyright-paperwork-exempt: yes` to the commit message.

**Comment by jorams** [comment]
I've amended the commit to fix the issues you mentioned. I do not have FSF papers signed.

Because the patch only returns `nil` if the actual value is `nil`, I don't think it causes issues if a non-boolean value is expected. The patch does also return `nil` if the value of the capability is an empty object. I am not 100% sure if that could be a problem.

**Comment by nemethf** [comment]
I looked all the eglot--server-capable calls, and they seems OK to me, even after this PR. 
(But there's #278, which is (I think) not a bug in eglot, but eglot could handle it more gracefully.) 

So I'm about to merge the PR in a day or two.

And this is the non-standard reply of felixfbecker/php-language-server:
```
(:result
 (:capabilities
  (:textDocumentSync 1 :hoverProvider t :completionProvider
                     (:resolveProvider :json-false :triggerCharacters
                                       ["$" ">"])
                     :signatureHelpProvider
                     (:triggerCharacters
                      ["(" ","])
                     :definitionProvider t :referencesProvider t :documentHighlightProvider nil :documentSymbolProvider t :workspaceSymbolProvider t :codeActionProvider nil :codeLensProvider nil :documentFormattingProvider nil :documentRangeFormattingProvider nil :documentOnTypeFormattingProvider nil :renameProvider nil :xworkspaceReferencesProvider t :xdefinitionProvider t :dependenciesProvider nil :xdependenciesProvider t))
 :id 1 :jsonrpc "2.0")
```

**Comment by joaotavora** [comment]
@nemethf OK.

**Comment by nemethf** [comment]
:( 

I was so confident that this patch was right, but it has turned out empty object might mean the corresponding capability is in fact supported.  I asked for clarification from microsoft/language-server-protocol.  

@jorams, can you list the all the servers you know sending nil instead of :json-false? It would be cleaner if the problem was solved their side after all.

**Comment by joaotavora** [comment]
> :(

Hey, don't fret. These things have subtleties.  But I've investigated this now:

* it seems that if a `nil` appears as the value to the key `:completionProvider` means one of two things:  an empty JSON object (`{}`) was transmitted or a `null` was transmitted.  Emacs's JSONRPC.el library interprets both these situations as `nil` (empty list) as appears to be good practice in [JSONland](https://stackoverflow.com/a/9623521/177259) but is different from the way that Javascript treats these values, for example.

* Additionally, Microsoft LSP's specification **seems** to specify that the way for the server to specify that it doesn't support a certain capability is to **omit** it from the capabilities list. 

  The  majority of servers seems to comply with this, which makes sense given the specific nature of LSP which works on top of capabilities (so old servers work with clients able to handle lots of capabilities and vice versa).  The servers that don't comply with this also do something reasonable. They put an explicit `false`in these fields, signalling something like "we know this exists but we don't support it".

  And then there's the php server which uses the ambiguous `null` instead of omitting or using `false`.

Now, the other issue, #285, where which was affected by this:

* It might be that the Elm server is sending the empty object there, interpreted as `nil` by `jsonrpc.el`, but as t by `eglot-server-capable`.
* It might be that it is sending `null`, also interpreted as `nil` by `jsonrpc.el` and as t by `eglot-server-capable`.

It could be argued that `jsonrpc.el` should change to distinguish these cases, namely always interpreting a `null` as am explicit `false`. But I'm far from 100% confident on that.  That would probably bring more breakage somewhere else, possibly massive breakage. It might be that in many places `null` is being transmitted to mean the empty object, and that would just break instantly.

So to summarize I think it is best to revert this change and the people that write the spec should be more careful about these subtleties. The author of the `php-server` should also be made aware of this.  Unfortunately I don't have time to engage neither of them right now.

**Comment by nemethf** [comment]
I reported the php server issue at felixfbecker/php-language-server#758 

> Microsoft LSP's specification **seems** to specify that the way for the server to specify that it doesn't support a certain capability is to omit it from the capabilities list.

This is now confirmed: microsoft/language-server-protocol#830

So I'm not against reverting this change, but it fixed other issues as well (I can only remember #278, or maybe that was the only other one).  Will 6dfdfff fix those other issues as well?

**Comment by joaotavora** [comment]
> So I'm not against reverting this change, but it fixed other issues as well (I can only remember #278, or maybe that was the only other one). Will 6dfdfff fix those other issues as well?

Yes it is quite probable it will. It seems to be the same problem from a brief analysis (expected a plist, but a non-plist found, and the only non-nil, non-plist I can think of is `t` and the `completionProvider: {}` case.

**Comment by nemethf** [comment]
@joaotavora, can you take a look at https://github.com/felixfbecker/php-language-server/issues/758?  It seems the bug might be on the client's side after all.

**Comment by joaotavora** [comment]
I don't understand this.  THere seems to be contradictory statements from the LSP maintainer: 

In https://github.com/microsoft/language-server-protocol/issues/830, @dbaeumer says:

> {"completionProvider": null} will not register a completion provider and is even invalid since null is not listed as a valid type."

@nemethf , can you clarify?

**Comment by dbaeumer** [comment]
The currrent VS Code client implementation is for historic reason a little fault tolerant for older properties. In general the spec specifies whether:

- it is optional
- can be null

Optional and null are two different things from the spec point of view.

**Comment by joaotavora** [comment]
@dbaeumer I've already asked you this in another form.  Your current statement seems to imply that the Php server is at fault and Eglot is behaving correctly,_though Eglot could be a little more fault tolerant in this specific case_. 

Do you know of any other servers besides Php's that are fault in this regard?

**Comment by dbaeumer** [comment]
Actually I am not tracking this so I don't know if there are other servers.

**Comment by joaotavora** [comment]
@dbaeumer thanks.  So I will assume for now that php is the only problem case.
